lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,106 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+static void generateMemberwiseInit(SourceEditConsumer &EditConsumer,
+                            SourceManager &SM,
+                            SmallVectorImpl<std::string>& memberNameVector,
+                            SmallVectorImpl<std::string>& memberTypeVector,
+                            SourceLoc targetLocation) {
+  
+  if (memberTypeVector.size() > 0) {
+    EditConsumer.accept(SM, targetLocation, "\ninternal init(");
+    
+    for (size_t i = 0; i < memberTypeVector.size(); i++) {
+      EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
+                          memberTypeVector[i]);
+      
+      if (i != memberTypeVector.size() - 1) {
+        EditConsumer.accept(SM, targetLocation, ", ");
+      }
+    }
+    
+    EditConsumer.accept(SM, targetLocation, ") {\n\n");
+    
+    for (auto varName: memberNameVector) {
+      EditConsumer.accept(SM, targetLocation,
+                          "self." + varName + " = " + varName + "\n");
+    }
+    
+    EditConsumer.accept(SM, targetLocation, "\n}\n");
+  }
+}
+  
+static SourceLoc collectMembersForInit(ResolvedCursorInfo CursorInfo,
+                           SmallVectorImpl<std::string>& memberNameVector,
+                           SmallVectorImpl<std::string>& memberTypeVector) {
+  
+  if (CursorInfo.Kind != CursorInfoKind::ValueRef && !CursorInfo.ValueD)
+    return SourceLoc();
+  
+  ClassDecl *classDecl = dyn_cast<ClassDecl>(CursorInfo.ValueD);
+  if (!classDecl || classDecl->getStoredProperties().empty() ||
+      CursorInfo.IsRef) {
+    return SourceLoc();
+  }
+  
+  SourceLoc bracesStart = classDecl->getBraces().Start;
+  if (!bracesStart.isValid())
+    return SourceLoc();
+  
+  SourceLoc targetLocation = bracesStart.getAdvancedLoc(1);
+  if (!targetLocation.isValid())
+    return SourceLoc();
+  
+  for (auto varDecl : classDecl->getStoredProperties()) {
+    auto parentPatternBinding = varDecl->getParentPatternBinding();
+    if (!parentPatternBinding)
+      continue;
+    
+    auto varDeclIndex =
+      parentPatternBinding->getPatternEntryIndexForVarDecl(varDecl);
+    
+    if (auto init = varDecl->getParentPatternBinding()->getInit(varDeclIndex)) {
+      if (init->getStartLoc().isValid())
+        continue;
+    }
+    
+    StringRef memberName = varDecl->getName().str();
+    memberNameVector.push_back(memberName.str());
+    
+    std::string memberType = varDecl->getType().getString();
+    memberTypeVector.push_back(memberType);
+  }
+  
+  return targetLocation;
```

Comment:
Can we ensure we return invalid `SourceLoc` if `memberNameVector` or `memberTypeVector` is empty? Currently, we may return valid location even if we don't need to insert anything.